File: sdks/java/src/dev/ironauth/verify/Fixtures.java
// SPDX-License-Identifier: MIT OR Apache-2.0
package dev.ironauth.verify;

import java.nio.charset.StandardCharsets;
import java.security.KeyPair;
import java.security.Signature;
import java.security.interfaces.EdECPublicKey;
import java.util.Base64;

/**
 * Minting helpers shared by {@link SelfTest} and {@link SampleHarness} (issue #118).
 *
 * <p>These exist because both suites need to sign a token the shared corpus could not contain:
 * the corpus has no private key, so anything requiring a VALID signature over attacker-chosen
 * headers has to be minted here.
 */
final class Fixtures {

    private Fixtures() {}

    /** Mint a genuinely signed Ed25519 JWS over the given header and claims, verbatim. */
    static String mint(KeyPair pair, String header, String claims) throws Exception {
        String signingInput =
                b64(header.getBytes(StandardCharsets.UTF_8)) + "." + b64(claims.getBytes(StandardCharsets.UTF_8));
        Signature signer = Signature.getInstance("Ed25519");
        signer.initSign(pair.getPrivate());
        signer.update(signingInput.getBytes(StandardCharsets.US_ASCII));
        return signingInput + "." + b64(signer.sign());
    }

    /**
     * The inverse of {@link TrustedKey}'s Ed25519 decoding: 32-byte little-endian y with the
     * parity of x in the top bit of the last byte (RFC 8032 5.1.2).
     */
    static String encodeEd25519(EdECPublicKey key) {
        byte[] bigEndian = key.getPoint().getY().toByteArray();
        byte[] raw = new byte[32];
        for (int i = 0; i < 32 && i < bigEndian.length; i++) {
            // toByteArray is big-endian and may carry a leading sign byte, so index from the END.
            raw[i] = bigEndian[bigEndian.length - 1 - i];
        }
        if (key.getPoint().isXOdd()) {
            raw[31] |= (byte) 0x80;
        }
        return b64(raw);
    }

    static String b64(byte[] bytes) {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
    }
}

File: sdks/dotnet/IronAuth.Verify.Checks/Fixtures.cs
// SPDX-License-Identifier: MIT OR Apache-2.0
using System.Text;
using IronAuth.Verify;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;
using Org.BouncyCastle.Security;

namespace IronAuth.Verify.Checks;

/// <summary>
/// Minting helpers shared by the self-test and the sample harness (issue #118).
/// </summary>
/// <remarks>
/// These exist because both suites need to sign a token the shared corpus could not contain: the
/// corpus carries no private key, so anything requiring a VALID signature over an attacker-chosen
/// header has to be minted here.
/// </remarks>
internal static class Fixtures
{
    /// <summary>Generate a fresh Ed25519 key pair.</summary>
    internal static AsymmetricCipherKeyPair GenerateEd25519()
    {
        Ed25519KeyPairGenerator generator = new();
        generator.Init(new Ed25519KeyGenerationParameters(new SecureRandom()));
        return generator.GenerateKeyPair();
    }

    /// <summary>Mint a genuinely signed Ed25519 JWS over the given header and claims, verbatim.</summary>
    internal static string Mint(AsymmetricCipherKeyPair pair, string header, string claims)
    {
        string signingInput = Base64Url.Encode(Encoding.UTF8.GetBytes(header))
            + "." + Base64Url.Encode(Encoding.UTF8.GetBytes(claims));
        byte[] bytes = Encoding.ASCII.GetBytes(signingInput);
        Ed25519Signer signer = new();
        signer.Init(true, pair.Private);
        signer.BlockUpdate(bytes, 0, bytes.Length);
        return signingInput + "." + Base64Url.Encode(signer.GenerateSignature());
    }

    /// <summary>The raw 32-byte public key, which is exactly what a JWK's <c>x</c> carries.</summary>
    internal static string PublicX(AsymmetricCipherKeyPair pair) =>
        Base64Url.Encode(((Ed25519PublicKeyParameters)pair.Public).GetEncoded());
}
